{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "data_path = 'data/HW4/'\n",
    "\n",
    "reads1 = f'{data_path}ecoli_400K_err_1.fastq'\n",
    "reads2 = f'{data_path}ecoli_400K_err_2.fastq'\n",
    "ref_genome = f'{data_path}MG1655-K12.first400K.fasta'\n",
    "\n",
    "spades_res = f'{data_path}spades_test/'\n",
    "spades_cor_path = f'{spades_res}/corrected/'\n",
    "spades_r_1 = f'{spades_cor_path}ecoli_400K_err_1.00.0_0.cor.fastq.gz'\n",
    "spades_r_2 = f'{spades_cor_path}ecoli_400K_err_2.00.0_0.cor.fastq.gz'\n",
    "\n",
    "bbduk_r_1 = f'{data_path}bbduk_tes_r_1p.fq'\n",
    "bbduk_r_2 = f'{data_path}bbduk_tes_r_2p.fq'\n",
    "bbduk_r_1u = f'{data_path}bbduk_tes_r_1u.fq'\n",
    "bbduk_r_2u = f'{data_path}bbduk_tes_r_2u.fq'\n",
    "\n",
    "aln = f'{data_path}test_aln.sam'\n",
    "aln_spades = f'{data_path}test_aln_spades.sam'\n",
    "aln_bbduk = f'{data_path}test_aln_bbduk.sam'\n",
    "\n",
    "aln_bam = f'{data_path}test_aln.bam'\n",
    "aln_spades_bam = f'{data_path}test_aln_spades.bam'\n",
    "aln_bbduk_bam = f'{data_path}test_aln_bbduk.bam'\n",
    "\n",
    "aln_bam_sorted = f'{data_path}test_aln.sorted.bam'\n",
    "aln_spades_bam_sorted = f'{data_path}test_aln_spades.sorted.bam'\n",
    "aln_bbduk_bam_sorted = f'{data_path}test_aln_bbduk.sorted.bam'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Исправление ошибок\n",
    "\n",
    "Для исправления ошибок будем использовать BayesHammer and Trimmomatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "== Warning ==  No assembly mode was specified! If you intend to assemble high-coverage multi-cell/isolate data, use '--isolate' option.\n",
      "\n",
      "\n",
      "Command line: /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/bin/spades.py\t-1\t/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_1.fastq\t-2\t/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_2.fastq\t--only-error-correction\t-o\t/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.15.2\n",
      "  Python version: 3.8.8\n",
      "  OS: macOS-10.15.4-x86_64-i386-64bit\n",
      "\n",
      "Output dir: /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test\n",
      "Mode: ONLY read error correction (without assembling)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Standard mode\n",
      "  For multi-cell/isolate data we recommend to use '--isolate' option; for single-cell MDA data use '--sc'; for metagenomic data use '--meta'; for RNA-Seq use '--rna'.\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_1.fastq']\n",
      "      right reads: ['/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_2.fastq']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Other parameters:\n",
      "  Dir for temp files: /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 250\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/spades.log\n",
      "\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running: /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/bin/spades-hammer /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     1M / 3M    INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from refs/heads/master, git revision 54bed604152fd271b68bdf7b78bfd0788ea9e431\n",
      "  0:00:00.001     1M / 3M    INFO    General                 (main.cpp                  :  76)   Loading config from /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/configs/config.info\n",
      "  0:00:00.003     1M / 3M    INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.003     1M / 3M    INFO    General                 (memory_limit.cpp          :  48)   Memory limit set to 250 Gb\n",
      "  0:00:00.003     1M / 3M    INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.004     1M / 3M    INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.004     1M / 3M    INFO    General                 (hammer_tools.cpp          :  38)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.004     1M / 3M    INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.005     1M / 3M    INFO    General                 (kmer_index_builder.hpp    : 243)   Splitting kmer instances into 16 files using 12 threads. This might take a while.\n",
      "  0:00:00.005     1M / 3M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 10240\n",
      "  0:00:00.005     1M / 3M    INFO    General                 (kmer_splitter.hpp         :  93)   Memory available for splitting buffers: 6.94444 Gb\n",
      "  0:00:00.005     1M / 3M    INFO    General                 (kmer_splitter.hpp         : 101)   Using cell size of 4194304\n",
      "  0:00:00.006  6913M / 6913M INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_1.fastq\n",
      "  0:00:05.684  6913M / 6913M INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1381602 reads\n",
      "  0:00:05.684  6913M / 6913M INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_2.fastq\n",
      "  0:00:10.929  6913M / 6913M INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 2763204 reads\n",
      "  0:00:10.929  6913M / 6913M INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 2763204 reads processed\n",
      "  0:00:10.929     1M / 21M   INFO    General                 (kmer_index_builder.hpp    : 249)   Starting k-mer counting.\n",
      "  0:00:11.357     1M / 99M   INFO    General                 (kmer_index_builder.hpp    : 260)   K-mer counting done. There are 13488952 kmers in total.\n",
      "  0:00:11.358     1M / 99M   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 395)   Building perfect hash indices\n",
      "  0:00:11.701    11M / 106M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 431)   Index built. Total 13488952 kmers, 9754400 bytes occupied (5.78512 bits per kmer).\n",
      "  0:00:11.701    11M / 106M  INFO   K-mer Counting           (kmer_data.cpp             : 354)   Arranging kmers in hash map order\n",
      "  0:00:12.030   219M / 312M  INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  0:00:22.885   219M / 312M  INFO    General                 (main.cpp                  : 155)   Extracting clusters:\n",
      "  0:00:22.886   219M / 312M  INFO    General                 (concurrent_dsu.cpp        :  18)   Connecting to root\n",
      "  0:00:22.944   219M / 312M  INFO    General                 (concurrent_dsu.cpp        :  34)   Calculating counts\n",
      "  0:00:23.555   288M / 380M  INFO    General                 (concurrent_dsu.cpp        :  63)   Writing down entries\n",
      "  0:00:26.045   219M / 312M  INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 2187352\n",
      "  0:00:26.052   115M / 209M  INFO   K-mer Counting           (kmer_data.cpp             : 371)   Collecting K-mer information, this takes a while.\n",
      "  0:00:26.196   427M / 518M  INFO   K-mer Counting           (kmer_data.cpp             : 377)   Processing /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_1.fastq\n",
      "  0:00:37.472   427M / 518M  INFO   K-mer Counting           (kmer_data.cpp             : 377)   Processing /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_2.fastq\n",
      "  0:00:48.543   427M / 518M  INFO   K-mer Counting           (kmer_data.cpp             : 384)   Collection done, postprocessing.\n",
      "  0:00:48.584   427M / 518M  INFO   K-mer Counting           (kmer_data.cpp             : 398)   There are 13488952 kmers in total. Among them 9831116 (72.8827%) are singletons.\n",
      "  0:00:48.584   427M / 518M  INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  0:00:50.379   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering done. Total 1 non-read kmers were generated.\n",
      "  0:00:50.379   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)   Subclustering statistics:\n",
      "  0:00:50.379   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton hamming clusters: 1393328. Among them 2580 (0.185168%) are good\n",
      "  0:00:50.379   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total singleton subclusters: 4144. Among them 4141 (99.9276%) are good\n",
      "  0:00:50.380   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Total non-singleton subcluster centers: 855746. Among them 836856 (97.7926%) are good\n",
      "  0:00:50.380   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average size of non-trivial subcluster: 14.1346 kmers\n",
      "  0:00:50.380   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Average number of sub-clusters per non-singleton cluster: 1.08295\n",
      "  0:00:50.380   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Total solid k-mers: 843577\n",
      "  0:00:50.380   427M / 545M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 658)     Substitution probabilities: [4,4]((0.925364,0.0417853,0.0246913,0.008159),(0.0168245,0.960387,0.012269,0.0105199),(0.0104931,0.0122445,0.960404,0.0168586),(0.00814172,0.0246661,0.0417493,0.925443))\n",
      "  0:00:50.381   427M / 529M  INFO    General                 (main.cpp                  : 178)   Finished clustering.\n",
      "  0:00:50.381   427M / 529M  INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 12 threads.\n",
      "  0:00:59.714   427M / 529M  INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 45120 new k-mers.\n",
      "  0:01:09.194   427M / 529M  INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 1693 new k-mers.\n",
      "  0:01:18.568   427M / 529M  INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 63 new k-mers.\n",
      "  0:01:28.094   427M / 529M  INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 3 produced 32 new k-mers.\n",
      "  0:01:37.611   427M / 529M  INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 4 produced 0 new k-mers.\n",
      "  0:01:37.611   427M / 529M  INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  0:01:37.611   427M / 529M  INFO    General                 (hammer_tools.cpp          : 222)   Starting read correction in 12 threads.\n",
      "  0:01:37.611   427M / 529M  INFO    General                 (hammer_tools.cpp          : 235)   Correcting pair of reads: /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_1.fastq and /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/ecoli_400K_err_2.fastq\n",
      "  0:01:39.671  1218M / 1273M INFO    General                 (hammer_tools.cpp          : 170)   Prepared batch 0 of 1200000 reads.\n",
      "  0:01:47.670  1258M / 1307M INFO    General                 (hammer_tools.cpp          : 177)   Processed batch 0\n",
      "  0:01:50.379  1258M / 1307M INFO    General                 (hammer_tools.cpp          : 187)   Written batch 0\n",
      "  0:01:50.688  1259M / 1307M INFO    General                 (hammer_tools.cpp          : 170)   Prepared batch 1 of 181602 reads.\n",
      "  0:01:51.953  1259M / 1307M INFO    General                 (hammer_tools.cpp          : 177)   Processed batch 1\n",
      "  0:01:52.381  1259M / 1307M INFO    General                 (hammer_tools.cpp          : 187)   Written batch 1\n",
      "  0:01:52.526   427M / 1106M INFO    General                 (hammer_tools.cpp          : 276)   Correction done. Changed 883777 bases in 651912 reads.\n",
      "  0:01:52.528   427M / 880M  INFO    General                 (hammer_tools.cpp          : 277)   Failed to correct 21932 bases out of 252262696.\n",
      "  0:01:52.565     1M / 468M  INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/corrected.yaml\n",
      "  0:01:52.565     1M / 468M  INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== corrected reads compression started. \n",
      "\n",
      "\n",
      "== Running: /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/bin/python /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/share/spades/spades_pipeline/scripts/compress_all.py --input_file /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/corrected.yaml --ext_python_modules_home /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/share/spades --max_threads 16 --output_dir /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected --gzip_output\n",
      "\n",
      "== Compressing corrected reads (with gzip)\n",
      "== Files to compress: ['/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/ecoli_400K_err_1.00.0_0.cor.fastq', '/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/ecoli_400K_err_2.00.0_0.cor.fastq', '/Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/ecoli_400K_err__unpaired.00.0_0.cor.fastq']\n",
      "== Files compression is finished\n",
      "== Dataset yaml file is updated\n",
      "\n",
      "===== corrected reads compression finished. \n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/bin/python /Users/nikitakonstantinovskiy/.conda/envs/HW_NGS/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/scaffolds.fasta --misc_dir /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Corrected reads are in /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/corrected/\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /Users/nikitakonstantinovskiy/Documents/Bioinformatics/IB/NGS/HW_NGS/data/HW4/spades_test/spades.log\n",
      "\n",
      "Thank you for using SPAdes!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run spades.py -1 {reads1} -2 {reads2} --only-error-correction -o {spades_res}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " -phred33 data/HW4/ecoli_400K_err_1.fastq data/HW4/ecoli_400K_err_2.fastq data/HW4/bbduk_tes_r_1p.fq data/HW4/bbduk_tes_r_1u.fq data/HW4/bbduk_tes_r_2p.fq data/HW4/bbduk_tes_r_2u.fq LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20\n",
      "Input Read Pairs: 1381602 Both Surviving: 1344536 (97,32%) Forward Only Surviving: 23598 (1,71%) Reverse Only Surviving: 10815 (0,78%) Dropped: 2653 (0,19%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "!java -jar utils/Trimmomatic-0.39/trimmomatic-0.39.jar PE -phred33 {reads1} {reads2} {bbduk_r_1} {bbduk_r_1u} \\\n",
    "{bbduk_r_2} {bbduk_r_2u} LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Выравнивание"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.00 sec\r\n",
      "[bwa_index] Construct BWT for the packed sequence...\r\n",
      "[bwa_index] 0.06 seconds elapse.\r\n",
      "[bwa_index] Update BWT... 0.00 sec\r\n",
      "[bwa_index] Pack forward-only FASTA... 0.00 sec\r\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.02 sec\r\n",
      "[main] Version: 0.7.17-r1188\r\n",
      "[main] CMD: bwa index data/HW4/MG1655-K12.first400K.fasta\r\n",
      "[main] Real time: 0.091 sec; CPU: 0.089 sec\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!conda run bwa index {ref_genome}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48301, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.315 CPU sec, 2.271 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48357, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.22, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.414 CPU sec, 2.279 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48345, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.499 CPU sec, 2.323 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48337, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.378 CPU sec, 2.237 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.403 CPU sec, 2.269 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48367, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.33, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.440 CPU sec, 2.291 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.505 CPU sec, 2.364 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48328, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.431 CPU sec, 2.293 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48359, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.447 CPU sec, 2.307 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48323, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.511 CPU sec, 2.372 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48276, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.532 CPU sec, 2.385 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48231, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.533 CPU sec, 2.384 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48187, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.564 CPU sec, 2.419 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48265, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.41, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.515 CPU sec, 2.368 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.401 CPU sec, 2.251 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.33, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.393 CPU sec, 2.245 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48303, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.379 CPU sec, 2.219 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48393, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.306 CPU sec, 2.162 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48320, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.320 CPU sec, 2.168 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.292 CPU sec, 2.146 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.314 CPU sec, 2.170 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48322, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.323 CPU sec, 2.169 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48361, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 100000 reads in 2.270 CPU sec, 2.127 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48350, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.308 CPU sec, 2.157 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48319, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.260 CPU sec, 2.108 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48334, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.282 CPU sec, 2.124 real sec\n",
      "[M::process] read 63204 sequences (6320400 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48289, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.238 CPU sec, 2.104 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 30551, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 63204 reads in 1.442 CPU sec, 1.333 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/HW4/MG1655-K12.first400K.fasta data/HW4/ecoli_400K_err_1.fastq data/HW4/ecoli_400K_err_2.fastq\n",
      "[main] Real time: 62.174 sec; CPU: 66.194 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48352, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 1.949 CPU sec, 1.767 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48401, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.22, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.044 CPU sec, 1.765 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48399, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.044 CPU sec, 1.770 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.020 CPU sec, 1.746 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.050 CPU sec, 1.771 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48414, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 100000 reads in 2.029 CPU sec, 1.749 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48317, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.062 CPU sec, 1.786 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48419, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.34, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.039 CPU sec, 1.757 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48404, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.048 CPU sec, 1.767 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.050 CPU sec, 1.767 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48318, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.046 CPU sec, 1.774 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.30, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.055 CPU sec, 1.781 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48260, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.098 CPU sec, 1.804 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48332, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.071 CPU sec, 1.777 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48379, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.072 CPU sec, 1.779 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48396, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.29, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.047 CPU sec, 1.765 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.45, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.052 CPU sec, 1.769 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48421, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.028 CPU sec, 1.746 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48381, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.065 CPU sec, 1.775 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48426, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.054 CPU sec, 1.773 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48336, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.024 CPU sec, 1.750 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48423, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.052 CPU sec, 1.770 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48404, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.049 CPU sec, 1.759 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48382, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.048 CPU sec, 1.758 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48355, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.024 CPU sec, 1.749 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48327, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.141 CPU sec, 1.860 real sec\n",
      "[M::process] read 16052 sequences (1605200 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 1.912 CPU sec, 1.762 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 7765, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 16052 reads in 0.405 CPU sec, 0.295 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/HW4/MG1655-K12.first400K.fasta data/HW4/spades_test//corrected/ecoli_400K_err_1.00.0_0.cor.fastq.gz data/HW4/spades_test//corrected/ecoli_400K_err_2.00.0_0.cor.fastq.gz\n",
      "[main] Real time: 48.290 sec; CPU: 55.784 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 110200 sequences (10000110 bp)...\n",
      "[M::process] read 110814 sequences (10000040 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53018, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110200 reads in 2.008 CPU sec, 1.959 real sec\n",
      "[M::process] read 111006 sequences (10000050 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53373, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110814 reads in 2.100 CPU sec, 1.960 real sec\n",
      "[M::process] read 111644 sequences (10000082 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53399, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.34, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111006 reads in 2.070 CPU sec, 1.924 real sec\n",
      "[M::process] read 112030 sequences (10000010 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53764, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.34, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111644 reads in 2.090 CPU sec, 1.929 real sec\n",
      "[M::process] read 113494 sequences (10000135 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53899, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 112030 reads in 2.138 CPU sec, 1.974 real sec\n",
      "[M::process] read 112048 sequences (10000183 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54584, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 113494 reads in 2.165 CPU sec, 2.000 real sec\n",
      "[M::process] read 112288 sequences (10000030 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53846, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.33, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 112048 reads in 2.134 CPU sec, 1.975 real sec\n",
      "[M::process] read 113242 sequences (10000098 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54064, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 112288 reads in 2.121 CPU sec, 1.966 real sec\n",
      "[M::process] read 113776 sequences (10000130 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54423, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 113242 reads in 2.154 CPU sec, 1.990 real sec\n",
      "[M::process] read 114870 sequences (10000069 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54646, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 113776 reads in 2.159 CPU sec, 1.991 real sec\n",
      "[M::process] read 114452 sequences (10000152 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 55107, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 114870 reads in 2.160 CPU sec, 1.998 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 110288 sequences (10000060 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54846, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 114452 reads in 2.203 CPU sec, 2.026 real sec\n",
      "[M::process] read 109970 sequences (10000159 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53098, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110288 reads in 2.151 CPU sec, 1.982 real sec\n",
      "[M::process] read 108832 sequences (10000054 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52915, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.30, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 109970 reads in 2.110 CPU sec, 1.939 real sec\n",
      "[M::process] read 108132 sequences (10000172 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52350, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.43, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 108832 reads in 2.083 CPU sec, 1.934 real sec\n",
      "[M::process] read 107444 sequences (10000045 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52116, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 108132 reads in 2.103 CPU sec, 1.938 real sec\n",
      "[M::process] read 107442 sequences (10000155 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51765, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107444 reads in 2.087 CPU sec, 1.931 real sec\n",
      "[M::process] read 107624 sequences (10000069 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51802, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.29, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107442 reads in 2.083 CPU sec, 1.919 real sec\n",
      "[M::process] read 107672 sequences (10000147 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51724, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107624 reads in 2.064 CPU sec, 1.913 real sec\n",
      "[M::process] read 107400 sequences (10000070 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51917, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107672 reads in 2.062 CPU sec, 1.902 real sec\n",
      "[M::process] read 106862 sequences (10000171 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51693, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107400 reads in 2.051 CPU sec, 1.894 real sec\n",
      "[M::process] read 106892 sequences (10000079 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51492, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.29, 10.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106862 reads in 2.058 CPU sec, 1.899 real sec\n",
      "[M::process] read 106614 sequences (10000185 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51426, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.22, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 106892 reads in 2.065 CPU sec, 1.903 real sec\n",
      "[M::process] read 44036 sequences (4155865 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51260, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106614 reads in 2.040 CPU sec, 1.909 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 21248, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.20, 10.24)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 44036 reads in 0.935 CPU sec, 0.809 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/HW4/MG1655-K12.first400K.fasta data/HW4/bbduk_tes_r_1p.fq data/HW4/bbduk_tes_r_2p.fq\n",
      "[main] Real time: 47.666 sec; CPU: 51.538 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa mem {ref_genome} {reads1} {reads2} > {aln}\n",
    "!bwa mem {ref_genome} {spades_r_1} {spades_r_2} > {aln_spades}\n",
    "!bwa mem {ref_genome} {bbduk_r_1} {bbduk_r_2} > {aln_bbduk}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2763204 + 0 in total (QC-passed reads + QC-failed reads)\r\n",
      "0 + 0 secondary\r\n",
      "0 + 0 supplementary\r\n",
      "0 + 0 duplicates\r\n",
      "2762617 + 0 mapped (99.98% : N/A)\r\n",
      "2763204 + 0 paired in sequencing\r\n",
      "1381602 + 0 read1\r\n",
      "1381602 + 0 read2\r\n",
      "2760820 + 0 properly paired (99.91% : N/A)\r\n",
      "2762132 + 0 with itself and mate mapped\r\n",
      "485 + 0 singletons (0.02% : N/A)\r\n",
      "0 + 0 with mate mapped to a different chr\r\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\r\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat {aln}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2716052 + 0 in total (QC-passed reads + QC-failed reads)\r\n",
      "0 + 0 secondary\r\n",
      "0 + 0 supplementary\r\n",
      "0 + 0 duplicates\r\n",
      "2715942 + 0 mapped (100.00% : N/A)\r\n",
      "2716052 + 0 paired in sequencing\r\n",
      "1358026 + 0 read1\r\n",
      "1358026 + 0 read2\r\n",
      "2714584 + 0 properly paired (99.95% : N/A)\r\n",
      "2715874 + 0 with itself and mate mapped\r\n",
      "68 + 0 singletons (0.00% : N/A)\r\n",
      "0 + 0 with mate mapped to a different chr\r\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\r\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat {aln_spades}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2689072 + 0 in total (QC-passed reads + QC-failed reads)\r\n",
      "0 + 0 secondary\r\n",
      "0 + 0 supplementary\r\n",
      "0 + 0 duplicates\r\n",
      "2687824 + 0 mapped (99.95% : N/A)\r\n",
      "2689072 + 0 paired in sequencing\r\n",
      "1344536 + 0 read1\r\n",
      "1344536 + 0 read2\r\n",
      "2685506 + 0 properly paired (99.87% : N/A)\r\n",
      "2686782 + 0 with itself and mate mapped\r\n",
      "1042 + 0 singletons (0.04% : N/A)\r\n",
      "0 + 0 with mate mapped to a different chr\r\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\r\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat {aln_bbduk}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\r\n"
     ]
    }
   ],
   "source": [
    "!samtools view -b {aln} > {aln_bam}\n",
    "!samtools sort {aln_bam} > {aln_bam_sorted}\n",
    "!samtools index {aln_bam_sorted}\n",
    "\n",
    "!samtools view -b {aln_spades} > {aln_spades_bam}\n",
    "!samtools sort {aln_spades_bam} > {aln_spades_bam_sorted}\n",
    "!samtools index {aln_spades_bam_sorted}\n",
    "\n",
    "!samtools view -b {aln_bbduk} > {aln_bbduk_bam}\n",
    "!samtools sort {aln_bbduk_bam} > {aln_bbduk_bam_sorted}\n",
    "!samtools index {aln_bbduk_bam_sorted}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Оценка качества исправления ошибок"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aln_err(line):\n",
    "    ref_seq = line.get_reference_sequence()\n",
    "    aln_seq = line.query_sequence\n",
    "    aln_seq = aln_seq[line.query_alignment_start: line.query_alignment_end]\n",
    "    r_ind = 0\n",
    "    a_ind = 0\n",
    "\n",
    "    err = []\n",
    "    \n",
    "    for ct in line.cigartuples:\n",
    "        if ct[0] == 0:\n",
    "            for i in range(ct[1]):\n",
    "                if ref_seq[r_ind].upper() != aln_seq[a_ind].upper():\n",
    "                    err.append(0)\n",
    "                else:\n",
    "                    err.append(1)\n",
    "                r_ind += 1\n",
    "                a_ind += 1\n",
    "        elif ct[0] == 1:\n",
    "            a_ind += ct[1]\n",
    "        elif ct[0] == 2:\n",
    "            r_ind += ct[1]\n",
    "    \n",
    "    return err\n",
    "\n",
    "def errors(orig_aln, err_cor_aln):\n",
    "    bamfile_orig = pysam.AlignmentFile(orig_aln, 'rb')\n",
    "    bamfile_err_cor = pysam.AlignmentFile(err_cor_aln, 'rb')\n",
    "    \n",
    "    ref_name_orig = bamfile_orig.get_reference_name(0)\n",
    "\n",
    "    bamfile_err_cor_ind = pysam.IndexedReads(bamfile_err_cor)\n",
    "    bamfile_err_cor_ind.build()\n",
    "    \n",
    "    errors = {\n",
    "        'Error in raw data': {'Error in corrected reads': 0, 'Correct base in corrected reads': 0, 'Base is absent in corrected reads': 0},\n",
    "        'Correct base in raw data': {'Error in corrected reads': 0, 'Correct base in corrected reads': 0, 'Base is absent in corrected reads': 0},\n",
    "    }\n",
    "        \n",
    "    for line in bamfile_orig.fetch(ref_name_orig):\n",
    "        read_type = line.is_read1\n",
    "        line_cor = None\n",
    "        try:\n",
    "            for lcor in bamfile_err_cor_ind.find(line.query_name):\n",
    "                if read_type == lcor.is_read1:\n",
    "                    line_cor = lcor\n",
    "                    break\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "        if line.cigartuples is not None:\n",
    "            err = aln_err(line)\n",
    "        else:\n",
    "            err = []\n",
    "        \n",
    "        if line_cor is not None and line_cor.cigartuples is not None:\n",
    "            err_cor = aln_err(line_cor)\n",
    "        else:\n",
    "            err_cor = []        \n",
    "        \n",
    "        i_err_cor = line.reference_start - line_cor.reference_start\n",
    "        \n",
    "        for i_err in range(len(err)):\n",
    "            raw_data = 'Error in raw data' if err[i_err] == 0 else 'Correct base in raw data'\n",
    "            cor_data = ''\n",
    "            \n",
    "            if i_err_cor < 0 or i_err_cor >= len(err_cor):\n",
    "                cor_data = 'Base is absent in corrected reads'\n",
    "            else:\n",
    "                cor_data = 'Error in corrected reads' if err_cor[i_err_cor] == 0 else 'Correct base in corrected reads'\n",
    "            \n",
    "            errors[raw_data][cor_data] += 1\n",
    "            \n",
    "            i_err_cor += 1\n",
    "        \n",
    "    return pd.DataFrame(errors).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BayesHammer Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Error in corrected reads</th>\n",
       "      <th>Correct base in corrected reads</th>\n",
       "      <th>Base is absent in corrected reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Error in raw data</th>\n",
       "      <td>247788</td>\n",
       "      <td>832542</td>\n",
       "      <td>1200515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Correct base in raw data</th>\n",
       "      <td>381100</td>\n",
       "      <td>246955705</td>\n",
       "      <td>16187708</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Error in corrected reads  \\\n",
       "Error in raw data                           247788   \n",
       "Correct base in raw data                    381100   \n",
       "\n",
       "                          Correct base in corrected reads  \\\n",
       "Error in raw data                                  832542   \n",
       "Correct base in raw data                        246955705   \n",
       "\n",
       "                          Base is absent in corrected reads  \n",
       "Error in raw data                                   1200515  \n",
       "Correct base in raw data                           16187708  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "errors(aln_bam_sorted, aln_spades_bam_sorted)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trimmomatic Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Error in corrected reads</th>\n",
       "      <th>Correct base in corrected reads</th>\n",
       "      <th>Base is absent in corrected reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Error in raw data</th>\n",
       "      <td>886504</td>\n",
       "      <td>1009</td>\n",
       "      <td>1344256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Correct base in raw data</th>\n",
       "      <td>614</td>\n",
       "      <td>241129878</td>\n",
       "      <td>19807902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Error in corrected reads  \\\n",
       "Error in raw data                           886504   \n",
       "Correct base in raw data                       614   \n",
       "\n",
       "                          Correct base in corrected reads  \\\n",
       "Error in raw data                                    1009   \n",
       "Correct base in raw data                        241129878   \n",
       "\n",
       "                          Base is absent in corrected reads  \n",
       "Error in raw data                                   1344256  \n",
       "Correct base in raw data                           19807902  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "errors(aln_bam_sorted, aln_bbduk_bam_sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
